File: ui/lib/analysis/organism_presets.dart
import 'dart:ui';

import 'package:collection/collection.dart';
import 'package:geneweb/genes/gene_list.dart';

class OrganismPresets {
  static final _arabidopsisStages = [
    // StageAndColor('tapetum_C', const Color(0xff993300)),
    // StageAndColor('EarlyPollen_C', const Color(0xffB71C1C), stroke: 4),
    // StageAndColor('UNM_C', const Color(0xffFF6D6D)),
    // StageAndColor('BCP_C', const Color(0xffC80002)),
    // StageAndColor('LatePollen_C', const Color(0xff0D47A1), stroke: 4),
    // StageAndColor('TCP_C', const Color(0xff21C5FF)),
    // StageAndColor('MPG_C', const Color(0xff305496)),
    // StageAndColor('SIV_C', const Color(0xffFF6600)),
    // StageAndColor('sperm_C', const Color(0xffFFC002)),
    // StageAndColor('leaves_C', const Color(0xff92D050)),
    // StageAndColor('seedling_C', const Color(0xffC6E0B4)),
    // StageAndColor('egg_C', const Color(0xff607D8B)),
    // StageAndColor('EarlyPollen_L', const Color(0xffB71C1C), stroke: 4),
    // StageAndColor('UNM_L', const Color(0xffFF6D6D)),
    // StageAndColor('BCP_L', const Color(0xffC80002)),
    // StageAndColor('LatePollen_L', const Color(0xff0D47A1), stroke: 4),
    // StageAndColor('TCP_L', const Color(0xff21C5FF)),
    // StageAndColor('MPG_L', const Color(0xff305496)),

    StageAndColor('col_tapetum', const Color(0xff993300)),
    StageAndColor('col_EarlyPollen', const Color(0xffB71C1C)),
    StageAndColor('col_UNM', const Color(0xffFF6D6D)),
    StageAndColor('col_BCP', const Color(0xffC80002)),
    StageAndColor('col_LatePollen', const Color(0xff0D47A1)),
    StageAndColor('col_TCP', const Color(0xff21C5FF)),
    StageAndColor('col_MPG', const Color(0xff305496)),
    StageAndColor('col_SIV_PT', const Color(0xffFF6600)),
    StageAndColor('col_sperm', const Color(0xffFFC002)),
    StageAndColor('col_leaves', const Color(0xff92D050)),
    StageAndColor('col_Seedlings', const Color(0xffC6E0B4)),
    StageAndColor('col_Egg', const Color(0xff607D8B)),

    StageAndColor('ler_EarlyPollen', const Color(0xffB71C1C)),
    StageAndColor('ler_UNM', const Color(0xffFF6D6D)),
    StageAndColor('ler_BCP', const Color(0xffC80002)),
    StageAndColor('ler_LatePollen', const Color(0xff0D47A1)),
    StageAndColor('ler_TCP', const Color(0xff21C5FF)),
    StageAndColor('ler_MPG', const Color(0xff305496)),

    // Chloroplast & Mitochondrion
    StageAndColor('Tapetum', const Color(0xff993300)),
    StageAndColor('EarlyPollen', const Color(0xffB71C1C)),
    StageAndColor('UNM', const Color(0xffFF6D6D)),
    StageAndColor('lerUNM', const Color(0xffFF6D6D)),
    StageAndColor('BCP', const Color(0xffC80002)),
    StageAndColor('lerBCP', const Color(0xffC80002)),
    StageAndColor('LatePollen', const Color(0xff0D47A1)),
    StageAndColor('TCP', const Color(0xff21C5FF)),
    StageAndColor('lerTCP', const Color(0xff21C5FF)),
    StageAndColor('MPG', const Color(0xff305496)),
    StageAndColor('lerMPG', const Color(0xff305496)),
    StageAndColor('SIV_PT', const Color(0xffFF6600)),
    StageAndColor('Sperm', const Color(0xffFFC002)),
    StageAndColor('Leaves', const Color(0xff92D050)),
    StageAndColor('Seedlings', const Color(0xffC6E0B4)),
    StageAndColor('Egg', const Color(0xff607D8B)),
  ];

  static Organism organismByFileName(String filename) {
    final preset = kOrganisms.firstWhereOrNull((element) => element.filename?.startsWith(filename) == true);
    if (preset != null) {
      return preset;
    }
    final name =
        RegExp(r'([A-Za-z0-9_]+).*').firstMatch(filename)?.group(1)?.replaceAll('_', ' ') ?? 'Unknown organism';
    return Organism(name: name, filename: filename);
  }

  static final List<Organism> kOrganisms = [
    Organism(
      name: 'Marchantia polymorpha',
      filename: 'Marchantia_polymorpha.fasta.zip',
      description: 'ATG',
      stages: [
        StageAndColor('Antheridium', const Color(0xff0085B4)),
        StageAndColor('Sperm_cell', const Color(0xffFFC002)),
        StageAndColor('Thallus', const Color(0xff548236)),
      ],
    ),
    Organism(
      public: true,
      name: 'Marchantia polymorpha',
      filename: 'Marchantia_polymorpha-with-tss.fasta.zip',
      description: 'ATG, TSS',
      stages: [
        StageAndColor('Antheridium', const Color(0xff0085B4)),
        StageAndColor('Sperm_cell', const Color(0xffFFC002)),
        StageAndColor('Thallus', const Color(0xff548236)),
      ],
    ),
    Organism(
      name: 'Physcomitrella patens',
      filename: 'Physcomitrella_patens.fasta.zip',
      description: 'ATG',
      stages: [
        StageAndColor('Antheridia_9DAI', const Color(0xff21C5FF)),
        StageAndColor('Antheridia_11DAI', const Color(0xff009ED6)),
        StageAndColor('14-15DAI_(mature)', const Color(0xff009AD0)),
        StageAndColor('Antheridia', const Color(0xff0085B4)),
        StageAndColor('Sperm_cell_packages', const Color(0xffFFDB69)),
        StageAndColor('Leaflets', const Color(0xff548236)),
      ],
    ),
    Organism(
      public: true,
      name: 'Physcomitrella patens',
      filename: 'Physcomitrella_patens-with-tss.fasta.zip',
      description: 'ATG, TSS',
      stages: [
        StageAndColor('Antheridia_9DAI', const Color(0xff21C5FF)),
        StageAndColor('Antheridia_11DAI', const Color(0xff009ED6)),
        StageAndColor('14-15DAI_(mature)', const Color(0xff009AD0)),
        StageAndColor('Antheridia', const Color(0xff0085B4)),
        StageAndColor('Sperm_cell_packages', const Color(0xffFFDB69)),
        StageAndColor('Leaflets', const Color(0xff548236)),
      ],
    ),
    Organism(
        public: true,
        name: 'Amborella trichopoda',
        filename: 'Amborella_trichopoda.fasta.zip',
        description: 'ATG',
        takeFirstTranscriptOnly: false,
        stages: [
          StageAndColor('UNM', const Color(0xffFF6D6D)),
          StageAndColor('Pollen', const Color(0xff0085B4)),
          StageAndColor('PT_bicellular', const Color(0xffE9A5D2)),
          StageAndColor('PT_tricellular', const Color(0xff77175C)),
          StageAndColor('Generative_cell', const Color(0xffB48502)),
          StageAndColor('Sperm_cell', const Color(0xffFFC002)),
          StageAndColor('Leaves', const Color(0xff92D050)),
        ]),
    Organism(
      public: true,
      name: 'Oryza sativa',
      filename: 'Oryza_sativa.fasta.zip',
      description: 'ATG',
      stages: [
        StageAndColor('TCP', const Color(0xff21C5FF)),
        StageAndColor('Pollen', const Color(0xff0085B4)),
        StageAndColor('Sperm', const Color(0xffFFC002)),
        StageAndColor('Leaves', const Color(0xff92D050)),
      ],
    ),
    Organism(
      name: 'Zea mays',
      filename: 'Zea_mays.fasta.zip',
      description: 'ATG',
      stages: [
        StageAndColor('Microspore', const Color(0xffFF6D6D)),
        //BCP missing
        StageAndColor('Pollen', const Color(0xff0085B4)),
        StageAndColor('PT', const Color(0xffE9A5D2)),
        StageAndColor('Sperm_cell', const Color(0xffFFC002)),
        StageAndColor('Leaves', const Color(0xff92D050)),
      ],
    ),
    Organism(
      public: true,
      name: 'Zea mays',
      filename: 'Zea_mays-with-tss.fasta.zip',
      description: 'ATG, TSS',
      stages: [
        StageAndColor('Microspore', const Color(0xffFF6D6D)),
        //BCP missing
        StageAndColor('Pollen', const Color(0xff0085B4)),
        StageAndColor('PT', const Color(0xffE9A5D2)),
        StageAndColor('Sperm_cell', const Color(0xffFFC002)),
        StageAndColor('Leaves', const Color(0xff92D050)),
      ],
    ),
    Organism(
      name: 'Solanum lycopersicum',
      filename: 'Solanum_lycopersicum.fasta.zip',
      description: 'ATG',
      stages: [
        StageAndColor('Microspore', const Color(0xffFF6D6D)),
        StageAndColor('Pollen', const Color(0xff0085B4)),
        StageAndColor('Pollen_grain', const Color(0xff305496)),
        StageAndColor('PT', const Color(0xffE9A5D2)),
        StageAndColor('PT_1,5h', const Color(0xffD75BAE)),
        StageAndColor('PT_3h', const Color(0xffAC2A81)),
        StageAndColor('PT_9h', const Color(0xff471234)),
        StageAndColor('Generative_cell', const Color(0xffB48502)),
        StageAndColor('Sperm_cell', const Color(0xffFFC002)),
        StageAndColor('Leaves', const Color(0xff92D050)),
      ],
    ),
    Organism(
      public: true,
      name: 'Solanum lycopersicum',
      filename: 'Solanum_lycopersicum-with-tss.fasta.zip',
      description: 'ATG, TSS',
      stages: [
        StageAndColor('Microspore', const Color(0xffFF6D6D)),
        StageAndColor('Pollen', const Color(0xff0085B4)),
        StageAndColor('Pollen_grain', const Color(0xff305496)),
        StageAndColor('PT', const Color(0xffE9A5D2)),
        StageAndColor('PT_1,5h', const Color(0xffD75BAE)),
        StageAndColor('PT_3h', const Color(0xffAC2A81)),
        StageAndColor('PT_9h', const Color(0xff471234)),
        StageAndColor('Generative_cell', const Color(0xffB48502)),
        StageAndColor('Sperm_cell', const Color(0xffFFC002)),
        StageAndColor('Leaves', const Color(0xff92D050)),
      ],
    ),
    Organism(
      name: 'Arabidopsis thaliana (ATG)',
      filename: 'Arabidopsis_thaliana.fasta.zip',
      description: 'ATG',
      stages: _arabidopsisStages,
    ),
    Organism(
      public: true,
      name: 'Arabidopsis thaliana (TSS)',
      filename: 'Arabidopsis_thaliana-with-tss.fasta.zip',
      description: 'ATG, TSS',
      stages: _arabidopsisStages,
    ),
    Organism(
      name: 'Arabidopsis thaliana',
      filename: 'Arabidopsis-variants.fasta.zip',
      description: 'TSS, ATG, all splicing variants',
      stages: _arabidopsisStages,
    ),
    Organism(
      name: 'Arabidopsis thaliana',
      filename: 'Arabidopsis_thaliana_mitochondrion.fasta.zip',
      description: 'Mitochondrion dataset',
      stages: _arabidopsisStages,
    ),
    Organism(
      name: 'Arabidopsis thaliana',
      filename: 'Arabidopsis_thaliana_chloroplast.fasta.zip',
      description: 'Chloroplast dataset',
      stages: _arabidopsisStages,
    ),
    Organism(
      name: 'Arabidopsis thaliana',
      filename: 'Arabidopsis_thaliana_small_rna.fasta.zip',
      description: 'Small RNA dataset',
      stages: [],
    ),
  ];
}

class Organism {
  final String name;
  final String? filename;
  final String? description;
  final bool public;
  final bool takeFirstTranscriptOnly;

  final List<StageAndColor> stages;

  Organism({
    required this.name,
    this.filename,
    this.description,
    this.public = false,
    this.takeFirstTranscriptOnly = true,
    this.stages = const [],
  });
}

File: backend/my_analysis_project/lib/analysis/organism_presets.py
import re
from organism import Organism
from stage_and_color import StageAndColor

class OrganismPresets:
    """
    Presets for organisms.

    See Organism
    """

    # Mimic the `_arabidopsisStages` list from Dart
    _arabidopsis_stages = [
        StageAndColor("C_Tapetum", "#993300"),
        StageAndColor("C_EarlyPollen", "#B71C1C"),
        StageAndColor("C_UNM", "#FF6D6D"),
        StageAndColor("C_BCP", "#C80002"),
        StageAndColor("C_LatePollen", "#0D47A1"),
        StageAndColor("C_TCP", "#21C5FF"),
        StageAndColor("C_MPG", "#305496"),
        StageAndColor("C_SIV_PT", "#FF6600"),
        StageAndColor("C_Sperm_cell", "#FFC002"),
        StageAndColor("C_Leaves_35d", "#92D050"),
        StageAndColor("C_Seedlings_10d", "#C6E0B4"),
        StageAndColor("C_Egg_cell", "#607D8B"),

        StageAndColor("L_EarlyPollen", "#B71C1C", is_checked_by_default=False),
        StageAndColor("L_UNM", "#FF6D6D", is_checked_by_default=False),
        StageAndColor("L_BCP", "#C80002", is_checked_by_default=False),
        StageAndColor("L_LatePollen", "#0D47A1", is_checked_by_default=False),
        StageAndColor("L_TCP", "#21C5FF", is_checked_by_default=False),
        StageAndColor("L_MPG", "#305496", is_checked_by_default=False),

        StageAndColor("Egg cell_Julca", "orange", is_checked_by_default=False),
        StageAndColor("Embryo", "orchid", is_checked_by_default=False),
        StageAndColor("Endosperm", "purple", is_checked_by_default=False),

        StageAndColor("Tapetum", "#993300"),
        StageAndColor("EarlyPollen", "#B71C1C"),
        StageAndColor("UNM", "#FF6D6D"),
        StageAndColor("lerUNM", "#FF6D6D"),
        StageAndColor("BCP", "#C80002"),
        StageAndColor("lerBCP", "#C80002"),
        StageAndColor("LatePollen", "#0D47A1"),
        StageAndColor("TCP", "#21C5FF"),
        StageAndColor("lerTCP", "#21C5FF"),
        StageAndColor("MPG", "#305496"),
        StageAndColor("lerMPG", "#305496"),
        StageAndColor("SIV_PT", "#FF6600"),
        StageAndColor("Sperm", "#FFC002"),
        StageAndColor("Leaves", "#92D050"),
        StageAndColor("Seedlings", "#C6E0B4"),
        StageAndColor("Egg", "#607D8B"),
    ]

    # Now define the master list `k_organisms`.
    # We'll match the same structure from Dart.
    k_organisms = [
        Organism(
            public=True,
            name="Marchantia polymorpha",
            filename="Marchantia_polymorpha-with-tss.fasta.zip",
            description="ATG, TSS",
            stages=[
                StageAndColor("Antheridium", "#0085B4"),
                StageAndColor("Sperm_cell", "#FFC002"),
                StageAndColor("Thallus", "#548236"),
            ],
        ),
        Organism(
            name="Marchantia polymorpha",
            filename="Marchantia_polymorpha.fasta.zip",
            description="ATG",
            stages=[
                StageAndColor("Antheridium", "#0085B4"),
                StageAndColor("Sperm_cell", "#FFC002"),
                StageAndColor("Thallus", "#548236"),
            ],
        ),
        Organism(
            name="Physcomitrium patens",
            filename="Physcomitrium_patens.fasta.zip",
            description="ATG",
            stages=[
                StageAndColor("Antheridia_9DAI", "#21C5FF"),
                StageAndColor("Antheridia_11DAI", "#009ED6"),
                StageAndColor("Antheridia_14-15DAI_(mature)", "#009AD0"),
                StageAndColor("Sperm_cell_packages", "#FFDB69"),
                StageAndColor("Leaflets", "#548236"),
                StageAndColor("Archegonia (mature)", "orange"),
                StageAndColor("Sporophyte (9 DAF)", "teal"),
            ],
        ),
        Organism(
            public=True,
            name="Physcomitrium patens",
            filename="Physcomitrium_patens-with-tss.fasta.zip",
            description="ATG, TSS",
            stages=[
                StageAndColor("Antheridia_9DAI", "#21C5FF"),
                StageAndColor("Antheridia_11DAI", "#009ED6"),
                StageAndColor("Antheridia_14-15DAI_(mature)", "#0085B4"),
                StageAndColor("Sperm_cell_packages", "#FFDB69"),
                StageAndColor("Leaflets", "#548236"),
                StageAndColor("Archegonia (mature)", "orange"),
                StageAndColor("Sporophyte (9 DAF)", "teal"),
            ],
        ),
        Organism(
            name="Azolla filiculoides",
            filename="Azolla_filiculoides.fasta.zip",
            description="ATG",
            stages=[
                StageAndColor("Leaves", "#92D050"),
                StageAndColor("Spores", "#FFC002"),
            ],
            take_first_transcript_only=False,
        ),
        Organism(
            name="Azolla filiculoides",
            filename="Azolla_filiculoides-with-tss.fasta.zip",
            description="ATG, TSS",
            stages=[
                StageAndColor("Leaves", "#92D050"),
                StageAndColor("Spores", "#FFC002"),
            ],
            take_first_transcript_only=False,
        ),
        Organism(
            name="Ceratopteris richardii",
            filename="Ceratopteris_richardii.fasta.zip",
            description="ATG",
            stages=[
                StageAndColor("Gametophyte", "#2980B9"),
                StageAndColor("Male_gametophyte", "#5DADE2"),
                StageAndColor("Hermaphrodite_gametophyte", "#8E44AD"),
                StageAndColor("Sporophyte", "#229954"),
            ],
        ),
        Organism(
            public=True,
            name="Ceratopteris richardii",
            filename="Ceratopteris_richardii-with-tss.fasta.zip",
            description="ATG, TSS",
            stages=[
                StageAndColor("Gametophyte", "#2980B9"),
                StageAndColor("Male_gametophyte", "#5DADE2"),
                StageAndColor("Hermaphrodite_gametophyte", "#8E44AD"),
                StageAndColor("Sporophyte", "#229954"),
            ],
        ),
        Organism(
            public=True,
            name="Amborella trichopoda",
            filename="Amborella_trichopoda.fasta.zip",
            description="ATG",
            take_first_transcript_only=False,
            stages=[
                StageAndColor("UNM", "#FF6D6D"),
                StageAndColor("Pollen", "#0085B4"),
                StageAndColor("PT_bicellular", "#E9A5D2"),
                StageAndColor("PT_tricellular", "#77175C"),
                StageAndColor("Generative_cell", "#B48502"),
                StageAndColor("Sperm_cell", "#FFC002"),
                StageAndColor("Leaves", "#92D050"),
            ],
        ),
        Organism(
            public=True,
            name="Oryza sativa",
            filename="Oryza_sativa.fasta.zip",
            description="ATG",
            stages=[
                StageAndColor("TCP", "#21C5FF"),
                StageAndColor("Pollen", "#0085B4"),
                StageAndColor("Sperm_cell", "#FFC002"),
                StageAndColor("Leaves", "#92D050"),
            ],
        ),
        Organism(
            name="Zea mays",
            filename="Zea_mays.fasta.zip",
            description="ATG",
            stages=[
                StageAndColor("Microspore", "#FF6D6D"),
                # BCP missing
                StageAndColor("Pollen", "#0085B4"),
                StageAndColor("PT", "#E9A5D2"),
                StageAndColor("Sperm_cell", "#FFC002"),
                StageAndColor("Leaves", "#92D050"),
            ],
        ),
        Organism(
            public=True,
            name="Zea mays",
            filename="Zea_mays-with-tss.fasta.zip",
            description="ATG, TSS",
            stages=[
                StageAndColor("Microspore", "#FF6D6D"),
                # BCP missing
                StageAndColor("Pollen", "#0085B4"),
                StageAndColor("PT", "#E9A5D2"),
                StageAndColor("Sperm_cell", "#FFC002"),
                StageAndColor("Leaves", "#92D050"),
            ],
        ),
        Organism(
            name="Solanum lycopersicum",
            filename="Solanum_lycopersicum.fasta.zip",
            description="ATG",
            stages=[
                StageAndColor("Microspore", "#FF6D6D"),
                StageAndColor("Pollen", "#0085B4"),
                StageAndColor("Pollen_grain", "#305496"),
                StageAndColor("PT", "#E9A5D2"),
                StageAndColor("PT_1,5h", "#D75BAE"),
                StageAndColor("PT_3h", "#AC2A81"),
                StageAndColor("PT_9h", "#471234"),
                StageAndColor("Generative_cell", "#B48502"),
                StageAndColor("Sperm_cell", "#FFC002"),
                StageAndColor("Leaves", "#92D050"),
            ],
        ),
        Organism(
            public=True,
            name="Solanum lycopersicum",
            filename="Solanum_lycopersicum-with-tss.fasta.zip",
            description="ATG, TSS",
            stages=[
                StageAndColor("Microspore", "#FF6D6D"),
                StageAndColor("Pollen", "#0085B4"),
                StageAndColor("Pollen_grain", "#305496"),
                StageAndColor("PT", "#E9A5D2"),
                StageAndColor("PT_1,5h", "#D75BAE"),
                StageAndColor("PT_3h", "#AC2A81"),
                StageAndColor("PT_9h", "#471234"),
                StageAndColor("Generative_cell", "#B48502"),
                StageAndColor("Sperm_cell", "#FFC002"),
                StageAndColor("Leaves", "#92D050"),
            ],
        ),
        Organism(
            name="Hordeum vulgare",
            filename="Hordeum_vulgare.fasta.zip",
            description="ATG",
            stages=[
                StageAndColor("Embryo_8_DAP", "#6E2C00"),
                StageAndColor("Embryo_16_DAP", "#A04000"),
                StageAndColor("Embryo_24_DAP", "#D35400"),
                StageAndColor("Embryo_32_DAP", "#E59866"),
                StageAndColor("Endosperm_4_DAP", "#7D6608"),
                StageAndColor("Endosperm_8_DAP", "#9A7D0A"),
                StageAndColor("Endosperm_16_DAP", "#B7950B"),
                StageAndColor("Endosperm_24_DAP", "#F1C40F"),
                StageAndColor("Endosperm_32_DAP", "#F7DC6F"),
                StageAndColor("Seed_maternal_tissues_4_DAP", "#1B4F72"),
                StageAndColor("Seed_maternal_tissues_8_DAP", "#2874A6"),
                StageAndColor("Seed_maternal_tissues_16_DAP", "#3498DB"),
                StageAndColor("Seed_maternal_tissues_24_DAP", "#85C1E9"),
                StageAndColor("Leaf_non-infested_30_DAP", "#82E0AA"),
            ],
        ),
        Organism(
            public=True,
            name="Hordeum vulgare",
            filename="Hordeum_vulgare-with-tss.fasta.zip",
            description="ATG, TSS",
            stages=[
                StageAndColor("Embryo_8_DAP", "#6E2C00"),
                StageAndColor("Embryo_16_DAP", "#A04000"),
                StageAndColor("Embryo_24_DAP", "#D35400"),
                StageAndColor("Embryo_32_DAP", "#E59866"),
                StageAndColor("Endosperm_4_DAP", "#7D6608"),
                StageAndColor("Endosperm_8_DAP", "#9A7D0A"),
                StageAndColor("Endosperm_16_DAP", "#B7950B"),
                StageAndColor("Endosperm_24_DAP", "#F1C40F"),
                StageAndColor("Endosperm_32_DAP", "#F7DC6F"),
                StageAndColor("Seed_maternal_tissues_4_DAP", "#1B4F72"),
                StageAndColor("Seed_maternal_tissues_8_DAP", "#2874A6"),
                StageAndColor("Seed_maternal_tissues_16_DAP", "#3498DB"),
                StageAndColor("Seed_maternal_tissues_24_DAP", "#85C1E9"),
                StageAndColor("Leaf_non-infested_30_DAP", "#82E0AA"),
            ],
        ),
        Organism(
            name="Arabidopsis thaliana",
            filename="Arabidopsis_thaliana.fasta.zip",
            description="ATG only",
            stages=_arabidopsis_stages,
        ),
        Organism(
            public=True,
            name="Arabidopsis thaliana",
            filename="Arabidopsis_thaliana-with-tss.fasta.zip",
            description="ATG, TSS",
            stages=_arabidopsis_stages,
        ),
        Organism(
            name="Arabidopsis thaliana",
            filename="Arabidopsis-variants.fasta.zip",
            description="TSS, ATG, all splicing variants",
            stages=_arabidopsis_stages,
        ),
        Organism(
            name="Arabidopsis thaliana",
            filename="Arabidopsis_thaliana_mitochondrion.fasta.zip",
            description="Mitochondrion dataset",
            stages=_arabidopsis_stages,
        ),
        Organism(
            name="Arabidopsis thaliana",
            filename="Arabidopsis_thaliana_chloroplast.fasta.zip",
            description="Chloroplast dataset",
            stages=_arabidopsis_stages,
        ),
        Organism(
            name="Arabidopsis thaliana",
            filename="Arabidopsis_thaliana_small_rna.fasta.zip",
            description="Small RNA dataset",
            stages=[],
        ),
        Organism(
            name="Allium cepa",
            filename="Allium_cepa.fasta.zip",
            description="ATG",
            stages=[],
        ),
        Organism(
            name="Silene vulgaris",
            filename="Silene_vulgaris.fasta.zip",
            description="ATG",
            stages=[],
        ),
        Organism(
            name="Silene vulgaris",
            filename="Silene_vulgaris-with-tss.fasta.zip",
            description="ATG, TSS",
            stages=[],
        ),
    ]

    @staticmethod
    def organism_by_filename(filename: str) -> "Organism":
        """
        Attempts to find an Organism in k_organisms whose filename starts with `filename`.
        If none is found, returns a new Organism with name based on the filename.
        """
        for org in OrganismPresets.k_organisms:
            if org.filename and org.filename.startswith(filename):
                return org

        # If not found, create a fallback organism
        match = re.match(r"([A-Za-z0-9_]+).*", filename)
        if match:
            fallback_name = match.group(1).replace("_", " ")
        else:
            fallback_name = "Unknown organism"

        return Organism(
            name=fallback_name,
            filename=filename,
        )
